aloneguid/parquet-dotnet
# [WIP] Modular Encryption Support When Reading Parquet Files

# Summary

I made significant progress on getting [Footer Decryption](https://github.com/apache/parquet-format/blob/master/Encryption.md#54-encrypted-footer-mode) working with parquet files (#191).

I'm opening this work-in-progress pull request with hopes that some other folks can help get this across the finish line. 

## AES_GCM_V1

Thanks to a test file file @pzatschl [shared with me](https://github.com/mukunku/ParquetViewer/discussions/92) I was able to implement the [Aes Gcm V1 encryption algorithm](https://github.com/apache/parquet-format/blob/master/Encryption.md#421-aes_gcm_v1). 

[link to code](https://github.com/mukunku/parquet-dotnet-DecryptFile_UTF8_AesGcmV1_192bit/blob/209c60f8cb7f3f8aa83cb9aef3da6f957d7fd5cd/src/Parquet/Encryption/AES_GCM_V1.cs#L32-L60)

## AES_GCM_CTR_V1

I also implemented the [Aes Gcm Ctr V1 encryption algorithm](https://github.com/apache/parquet-format/blob/master/Encryption.md#422-aes_gcm_ctr_v1), however I don't have any test files to confirm it's working 🙃 

[link to code](https://github.com/mukunku/parquet-dotnet-DecryptFile_UTF8_AesGcmV1_192bit/blob/209c60f8cb7f3f8aa83cb9aef3da6f957d7fd5cd/src/Parquet/Encryption/AES_GCM_CTR_V1.cs#L28-L71)

# How to test

Checkout the unit test I added that tests the sample file I mentioned above:
[link to code](https://github.com/mukunku/parquet-dotnet-DecryptFile_UTF8_AesGcmV1_192bit/blob/873d528d2d0d1cf5e55978ed4491eb0bf077a072/src/Parquet.Test/ParquetReaderOnTestFilesTest.cs#L177-L192)
![image](https://github.com/aloneguid/parquet-dotnet/assets/4502154/03bd0ea9-6fb2-42e2-8b46-ad832314b1f0)



However, even though I can decrypt the test file successfully, the data itself doesn't seem to be valid. So I had to add this try-catch as a temporary workaround. 
[link to code](https://github.com/mukunku/parquet-dotnet-DecryptFile_UTF8_AesGcmV1_192bit/blob/873d528d2d0d1cf5e55978ed4491eb0bf077a072/src/Parquet/Extensions/OtherExtensions.cs#L16-L23)
We should remove this once we have a proper test file. (Unfortunately I don't have any other test files )
